File: app/src/main/java/quintype/com/templatecollectionwithrx/models/story/StoryElementSubTypeMetaData.java
package quintype.com.templatecollectionwithrx.models.story;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.StringDef;
import android.text.TextUtils;

import com.google.gson.annotations.SerializedName;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * An object instance representing StoryElementSubTypeMetaData
 *
 * @author Imran imran@quintype.com
 * @author Madhu madhu@quintype.com
 */

public class StoryElementSubTypeMetaData implements Parcelable {

    @Retention(RetentionPolicy.SOURCE)
    @StringDef({TYPE_SLIDESHOW, TYPE_GALLERY, TYPE_INVALID})
    public @interface Type {
    }


    public static final String TYPE_SLIDESHOW = "slideshow";
    public static final String TYPE_GALLERY = "gallery";
    public static final String TYPE_INVALID = "invalid";

    @SerializedName("tweet-url")
    private String tweetUrl;
    @SerializedName("tweet-id")
    private String tweetId;
    @SerializedName("attribution")
    private String attribution;
    @SerializedName("content")
    private String content;
    @SerializedName("type")
    private String type;
    @SerializedName("question")
    private String question;
    @SerializedName("answer")
    private String answer;
    @SerializedName("video-id")
    public String videoId;
    @SerializedName("file-name")
    String fileName;
    @SerializedName("has-header")
    boolean hasHeader;
    @SerializedName("linked-story")
    LinkedStory linkedStory;
    @SerializedName("linked-story-id")
    public String linkedStoryId;
    @SerializedName("account-id")
    private String mBrightCoveAccountID;
    @SerializedName("poster-url")
    private String mBrightCovePosterURL;

    public boolean hasHeader() {
        return hasHeader;
    }

    @Override
    public String toString() {
        return "StoryElementSubTypeMetaData{" +
                "tweetUrl='" + tweetUrl + '\'' +
                ", tweetId='" + tweetId + '\'' +
                ", attribution='" + attribution + '\'' +
                ", content='" + content + '\'' +
                ", type='" + type + '\'' +
                ", question='" + question + '\'' +
                ", answer='" + answer + '\'' +
                ", videoId='" + videoId + '\'' +
                ", hasHeader='" + hasHeader + '\'' +
                ", linkedStory='" + linkedStory + '\'' +
                ", linkedStoryId='" + linkedStoryId + '\'' +
                ", mBrightCoveAccountID='" + mBrightCoveAccountID + '\'' +
                ", mBrightCovePosterURL='" + mBrightCovePosterURL + '\'' +
                '}';
    }


    /**
     * Tweet Url
     *
     * @return String tweet url
     */
    public String tweetUrl() {
        return tweetUrl;
    }

    /**
     * Tweet Url
     *
     * @return String video Id
     */
    public String videoId() {
        return videoId;
    }

    /**
     * Tweet Id
     *
     * @return String tweet id
     */
    public String tweetId() {
        return tweetId;
    }

    /**
     * @return String attribution
     */
    public String attribution() {
        return attribution;
    }

    /**
     * @return String content
     */
    public String content() {
        return content;
    }

    /**
     * Type of SubElement
     *
     * @return String {@link Type}
     */
//    public
//    @Type
//    String type() {
//        if (TextUtils.isEmpty(type)) {
//            return TYPE_INVALID;
//        }
//
//        if (isTypeSlideShow()) {
//            return TYPE_SLIDESHOW;
//        } else if (isTypeGallery()) {
//            return TYPE_GALLERY;
//        }
//        return TYPE_INVALID;
//    }

    /**
     * Check if SubElement is of type gallery
     *
     * @return true if the element is gallery else false
     */
//    public boolean isTypeGallery() {
//        return StringUtils.equalsIgnoreCase(type, TYPE_GALLERY);
//    }
//
//    /**
//     * Check if SubElement is of type slide show
//     *
//     * @return true is the element is slideshow else false
//     */
//    public boolean isTypeSlideShow() {
//        return StringUtils.equalsIgnoreCase(type, TYPE_SLIDESHOW);
//    }

    public String question() {
        return question;
    }

    public String answer() {
        return answer;
    }

    public StoryElementSubTypeMetaData() {
    }

    public String getmBrightCoveAccountID() {
        return mBrightCoveAccountID;
    }

    public String getVideoId() {
        return videoId;
    }

    public String getLinkedStoryId() {
        return linkedStoryId;
    }


    public String getmBrightCovePosterURL() {
        return mBrightCovePosterURL;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.tweetUrl);
        dest.writeString(this.tweetId);
        dest.writeString(this.attribution);
        dest.writeString(this.content);
        dest.writeString(this.type);
        dest.writeString(this.question);
        dest.writeString(this.answer);
        dest.writeString(this.videoId);
        dest.writeByte(hasHeader ? (byte) 1 : (byte) 0);
        dest.writeParcelable(this.linkedStory, 0);
        dest.writeString(this.linkedStoryId);
        dest.writeString(this.mBrightCoveAccountID);
        dest.writeString(this.mBrightCovePosterURL);
    }

    protected StoryElementSubTypeMetaData(Parcel in) {
        this.tweetUrl = in.readString();
        this.tweetId = in.readString();
        this.attribution = in.readString();
        this.content = in.readString();
        this.type = in.readString();
        this.question = in.readString();
        this.answer = in.readString();
        this.videoId = in.readString();
        this.hasHeader = in.readByte() != 0;
        this.linkedStory = in.readParcelable(LinkedStory.class.getClassLoader());
        this.linkedStoryId = in.readString();
        this.mBrightCoveAccountID = in.readString();
        this.mBrightCovePosterURL = in.readString();
    }

    public static final Creator<StoryElementSubTypeMetaData> CREATOR = new Creator<StoryElementSubTypeMetaData>() {
        public StoryElementSubTypeMetaData createFromParcel(Parcel source) {
            return new StoryElementSubTypeMetaData(source);
        }

        public StoryElementSubTypeMetaData[] newArray(int size) {
            return new StoryElementSubTypeMetaData[size];
        }
    };
}

File: app/src/main/java/quintype/com/templatecollectionwithrx/models/story/StoryElementSubTypeMetaData.kt
package quintype.com.templatecollectionwithrx.models.story

import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.StringDef
import android.text.TextUtils
import com.google.gson.annotations.SerializedName
import quintype.com.templatecollectionwithrx.utils.Constants
import quintype.com.templatecollectionwithrx.utils.Constants.Companion.TYPE_GALLERY
import quintype.com.templatecollectionwithrx.utils.Constants.Companion.TYPE_INVALID
import quintype.com.templatecollectionwithrx.utils.Constants.Companion.TYPE_SLIDESHOW
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy

/**
 * Created TemplateCollectionWithRx by rakshith on 9/4/18.
 */

class StoryElementSubTypeMetaData : Parcelable {

    @SerializedName("tweet-url")
    private val tweetUrl: String
    @SerializedName("tweet-id")
    public val tweetId: String
    @SerializedName("attribution")
    private val attribution: String
    @SerializedName("content")
    private val content: String
    @SerializedName("type")
    private val type: String
    @SerializedName("question")
    private val question: String
    @SerializedName("answer")
    private val answer: String
    @SerializedName("video-id")
    var videoId: String
    @SerializedName("file-name")
    internal var fileName: String? = null
    @SerializedName("has-header")
    internal var hasHeader: Boolean = false
    @SerializedName("linked-story")
    internal var linkedStory: LinkedStory
    @SerializedName("linked-story-id")
    var linkedStoryId: String
    @SerializedName("account-id")
    private val mBrightCoveAccountID: String
    @SerializedName("poster-url")
    private val mBrightCovePosterURL: String

    @Retention(RetentionPolicy.SOURCE)
    @StringDef(Constants.TYPE_SLIDESHOW, Constants.TYPE_GALLERY, Constants.TYPE_INVALID)
    annotation class Type

    override fun toString(): String {
        return "StoryElementSubTypeMetaData{" +
                "tweetUrl='" + tweetUrl + '\''.toString() +
                ", tweetId='" + tweetId + '\''.toString() +
                ", attribution='" + attribution + '\''.toString() +
                ", content='" + content + '\''.toString() +
                ", type='" + type + '\''.toString() +
                ", question='" + question + '\''.toString() +
                ", answer='" + answer + '\''.toString() +
                ", videoId='" + videoId + '\''.toString() +
                ", hasHeader='" + hasHeader + '\''.toString() +
                ", linkedStory='" + linkedStory + '\''.toString() +
                ", linkedStoryId='" + linkedStoryId + '\''.toString() +
                ", mBrightCoveAccountID='" + mBrightCoveAccountID + '\''.toString() +
                ", mBrightCovePosterURL='" + mBrightCovePosterURL + '\''.toString() +
                '}'.toString()
    }

    /**
     * Type of SubElement
     *
     * @return String [Type]
     */
    fun type(): String {
        return if (TextUtils.isEmpty(this.type)) {
            "invalid"
        } else if (this.isTypeSlideShow()) {
            "slideshow"
        } else {
            if (this.isTypeGallery()) "gallery" else "invalid"
        }
    }

    /**
     * Check if SubElement is of type gallery
     *
     * @return true if the element is gallery else false
     */
    fun isTypeGallery(): Boolean {
        return type.equals(TYPE_GALLERY, true)
    }

    /**
     * Check if SubElement is of type slide show
     *
     * @return true is the element is slideshow else false
     */
    fun isTypeSlideShow(): Boolean {
        return type.equals(TYPE_SLIDESHOW, true)
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.tweetUrl)
        dest.writeString(this.tweetId)
        dest.writeString(this.attribution)
        dest.writeString(this.content)
        dest.writeString(this.type)
        dest.writeString(this.question)
        dest.writeString(this.answer)
        dest.writeString(this.videoId)
        dest.writeByte(if (hasHeader) 1.toByte() else 0.toByte())
        dest.writeParcelable(this.linkedStory, 0)
        dest.writeString(this.linkedStoryId)
        dest.writeString(this.mBrightCoveAccountID)
        dest.writeString(this.mBrightCovePosterURL)
    }

    protected constructor(parcel: Parcel) {
        this.tweetUrl = parcel.readString()
        this.tweetId = parcel.readString()
        this.attribution = parcel.readString()
        this.content = parcel.readString()
        this.type = parcel.readString()
        this.question = parcel.readString()
        this.answer = parcel.readString()
        this.videoId = parcel.readString()
        this.hasHeader = parcel.readByte().toInt() != 0
        this.linkedStory = parcel.readParcelable(LinkedStory::class.java.classLoader)
        this.linkedStoryId = parcel.readString()
        this.mBrightCoveAccountID = parcel.readString()
        this.mBrightCovePosterURL = parcel.readString()
    }

    companion object CREATOR : Parcelable.Creator<StoryElementSubTypeMetaData> {
        override fun createFromParcel(parcel: Parcel): StoryElementSubTypeMetaData {
            return StoryElementSubTypeMetaData(parcel)
        }

        override fun newArray(size: Int): Array<StoryElementSubTypeMetaData?> {
            return arrayOfNulls(size)
        }
    }
}